aiida_quantumespresso/calculations/base.py

```diff
@@ -0,0 +1,48 @@
+# -*- coding: utf-8 -*-
+"""Defines a CalcJob base class for aiida-quantumespresso.
+
+The custom CalcJob base class automatically sets the'invalidates_cache'
+attribute of exit codes based on the status integer.
+"""
+
+from __future__ import absolute_import
+
+from aiida.engine import ExitCode, CalcJob as _BaseCalcJob
+from aiida.engine.processes.process_spec import CalcJobProcessSpec as _BaseCalcJobProcessSpec
+
+__all__ = ('CalcJob',)
+
+
+class CalcJobProcessSpec(_BaseCalcJobProcessSpec):
+    """Process spec for aiida-quantumespresso CalcJob classes.
+
+    Automatically sets the `invalidates_cache` flag to `True` for an
+    exit status smaller than 400, unless explicitly overriden.
+    """
+
+    def exit_code(self, status, label, message, invalidates_cache=None):
+        """Add an exit code to the ProcessSpec.
+
+        .. note: for any status smaller than `400` the `invalidates_cache` is
+            automatically set to `True`, unless explicitly overridden.
+
+        :param status: the exit status integer
+        :param label: a label by which the exit code can be addressed
+        :param message: a more detailed description of the exit code
+        :param invalidates_cache: when set to `True`, a process exiting
+            with this exit code will not be considered for caching
+        """
+        if invalidates_cache is None:
+            invalidates_cache = (isinstance(status, int) and status < 400)
+
+        if 'invalidates_cache' in ExitCode._fields:
```

Comment:
Another unforeseen usage of `ExitCode` that would break with my [proposed change in `aiida-core](https://github.com/aiidateam/aiida-core/pull/3824) :)  can we turn this into a try-catch of `AttributeError`? That should work even if we decide to move `ExitCode` to a simple class, right?

Reply:
Good point - it should be a catch of `TypeError`  when instantiating the `ExitCode` however, because we're checking on the class itself here, not the instance.